packages/providers/src/provider.ts

```diff
@@ -595,12 +603,27 @@ export default class Provider {
       );
     }
 
+    const encodedTransaction = hexlify(transactionRequest.toTransactionBytes());
+
+    if (awaitExecution) {
+      const subscription = this.operations.submitAndAwait({ encodedTransaction });
+      for await (const { submitAndAwait } of subscription) {
+        if (submitAndAwait.type === 'SuccessStatus') {
+          break;
+        }
+      }
```

Comment:
What happens if the response type is different from `SuccessStatus` ?

Reply:
Good question. By default, the fuel node closes the connection when it returns a non-`SubmittedStatus` after executing/squeezing out the transaction. This, in turn, causes the `for-await-of` loop to close as well.

The code itself currently doesn't handle any non-successful transactions. In theory, we should throw errors to the user if a non-`SuccessStatus` status is returned. In practice, however, this needs to be fleshed out and centralized because currently there are multiple places where statuses/failure receipts are handled ([1](https://github.com/FuelLabs/fuels-ts/blob/78e5e405ef442b47032943196f67c3c0db125c92/packages/providers/src/transaction-response/transaction-response.ts#L240), [2](https://github.com/FuelLabs/fuels-ts/blob/78e5e405ef442b47032943196f67c3c0db125c92/packages/program/src/errors.ts#L34), and possibly more). I'm mentioning failure receipts here alongside different statuses because of #1588. I already tried handling `FailureStatus` and `SqueezedOutStatus` but realized that it deserves a different PR because it's not trivial given the codebase handles them already in other places via the `TransactionResponse`'s transaction summary.

Reply:
Okay, so it was weird that the `for-await-of` wasn't ending because, as it says [here](https://github.com/FuelLabs/fuels-ts/blob/d3a0cffb29d371218375df5fb26d2b75982d33f2/packages/providers/fuel-core-schema.graphql#L1003), the connection does close after a non-`SubmittedStatus` is sent. I even remember seeing `done` be set to `true` [here](https://github.com/FuelLabs/fuels-ts/blob/ad5923d40fdcfcd8b8e7f4473449a7b28981d715/packages/providers/src/fuel-graphql-subscriber.ts#L77), where `done` is the indicator that the stream ended, so I assumed everything worked. Anyways, after I moved from piping `ReadableStream` to using `for-await-of` internally in `fuel-graphql-subscriber.ts` (https://github.com/FuelLabs/fuels-ts/pull/1615/commits/bb9e117dd47968b7775f82cf2cb884454f718900), the stream closes properly and as expected. This was a bug that got fixed by that commit.

Reply:
Unfortunately I had to revert the above commit in https://github.com/FuelLabs/fuels-ts/pull/1615/commits/01267c0373007bfba09382f9c3331c4e0568ac10 because `for-await-of` is not supported for `ReadableStream` in most browsers ([MDN](https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream#browser_compatibility), see the `@@asyncIterator` row). However, I added a fix to at least close the stream intentionally in https://github.com/FuelLabs/fuels-ts/pull/1615/commits/9b3a05706df334115a23e45ed74333824da12a41 by refactoring from an async generator function to crafting our own async iterator. I assume the `ReadableStream` was closed in the previous implementation as well via garbage collection, but now we're being explicit about it. The `return` method, in which the stream is closed, is called when a `break` is hit inside a `for-await-of` loop.

I created #1628 to further investigate this issue. It isn't a blocker though, because we're breaking the `for-await-of` loops ourselves and aren't relying on connections closing for any behavior.

Reply:
Phew, the whole issue is fixed in https://github.com/FuelLabs/fuels-ts/pull/1615/commits/7ff5d1637871f6dc5ebf93cc771c998eed668227. The custom stream transformer `FuelSubscriptionStream` had an issue. After replacing it with the `parseBytesStream` everything works.

Reply:
@nedsalk Sorry but I am having a hard time understanding why this problem happened in the first place and how you managed to solve it. Can you help me understand?

Reply:
When using GraphQL subscriptions, we are using the [Streams API](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API) to read the streams of data being sent from the fuel node via an SSE connection. The fuel node guarantees that it'll close this SSE connection after the status of a transaction changes from `SubmittedStatus` to any other. On our side, the indicator that the connection closed which would, in turn, close our stream was being lost inside of the `FuelSubscriptionStream` class that was transforming the stream of data being sent from the fuel node. This caused the [loop that was expecting data from the server](https://github.com/FuelLabs/fuels-ts/blob/ad5923d40fdcfcd8b8e7f4473449a7b28981d715/packages/providers/src/fuel-graphql-subscriber.ts#L71) to loop to infinity because `done` (the indicator of the stream ending) was never `true`. This slipped through the cracks because we'd always manually close the whole async generator with a `break` of the `for-await-of` loops that are consuming the generator.

Everything now works after I removed this `FuelSubscriptionStream` transformer and opted for a simple function that wasn't transforming the actual stream read from `response.body` but just parsing its bytes. After the connection is closed by the fuel node, that stream returns `{done: true, value: undefined}` and the loop inside of the async generator will break on the `if(done) break` condition and the generator will finish.

Reply:
I reverted this bugfix (https://github.com/FuelLabs/fuels-ts/pull/1615/commits/ca6b64331e050052815c85da0efaf24f3f854dd6) and moved it to #1597 because the bug presented itself there as well when I was writing tests for the subscriptions and that PR is more appropriate. I don't want to have two PRs that are solving the same issue open at the same time.